File: solutions/esab-atad/js/ESAb_ATAd.js
/***
 * EncorAcademy 2020B
 * ESAb ATAd
 * Google Code Jam 2020 
 * 
 * */
const rl = require('readline').createInterface({
    input: process.stdin,
    output: process.stdout
});

function flip(arr){
  for(let i = 0; i < arr.length; i++)
  {
      arr[i] ^= 1;
  }
  return arr;
}

async function query(i, count) {
    if(i !== null)
    {
        console.log(i+1);
    }
    else
    {
        console.log(1);
    }
    rl.resume();
    let input = await getInput();
    rl.pause();
    return [Number(input), count+1];
}

async function check(result){
    // unir arreglo de resultado en un string
    const res = result.join('');
    // recibe input del juez
    // ingresa "Y" cuando el resultado es correcto
    // ingresa "N" cuando el resultado es incorrecto
    console.log(res);
    rl.resume();
    const ok = await getInput();
    rl.pause();
    // rl.close();
    //verifica que el resutado sea correcto
    if (ok != "Y") // error
    {
        process.exit(1);
    }
}

async function esab_atad(B)
{
    //result arreglo con la longitud de B
    //count contador de el numero de queries realizados
    let result = Array.from({length: B}, (_, i) => 0);
    let count = 0;
    let flip_i = null;
    let reverse_i = null;

    let lengthFloor = Math.floor(B/2);
    let flip_res = null;
    let reverse_res = null;

    //para cada elemento del numero de Bits / 2 (redondeado hacia abajo)
    for(let i = 0; i < lengthFloor; i++){
        // console.log(i);
        if (count%10 === 0 && i!=0){
            // recibe el flag del tester y count + 1 y los guardar en flip_res y count
            [flip_res, count] = await query(flip_i, count);
            //recibe el flag del tester y count + 1 y los guardar en reverse_res y count
            [reverse_res, count] = await query(reverse_i, count)
            if (flip_i !== null && (result[flip_i] ^ flip_res)){
               //manda llamar la funcion flip y modifica todo el arreglo
               result = flip(result);
            }
            if (reverse_i !== null && (result[reverse_i] ^ reverse_res))
            {
                //invierte el arreglo
                result = result.reverse();
            }
                
        }
        // Regresa el valor que contiene el input original en la posicion i
        // y lo asigna en la posicion i del arreglo result
        [result[i], count] = await query(i, count);
        // Regresa el valor que contiene el input original en la posicion B-i-1
        // y lo asigna en la posicion B-i-1 del arreglo result
        [result[B-i-1], count] = await query(B-i-1, count);
        if (result[i] === result[B-i-1])
        {
            flip_i = i;
        }
        else
        {
            reverse_i = i;
        }
    }
    await check(result);
}

function getInput()
{
    return new Promise((resolve, reject) => {
      rl.once("line", (answer) => {
        resolve(answer)
      })
    })
}
  

async function main() {
    const res = await getInput();

    const T = res.split(' ')[0];
    const B = res.split(' ')[1];
    rl.pause();
    for (let test = 0; test<T;test++)
    {
        await esab_atad(B);
    }
    process.exit(0);
}
main()

File: solutions/esab-atad/java/esab-atad.java
import java.util.Scanner;

class Solution {

  	public static Scanner sc= new Scanner(System.in);
  	public static int count = 0;

  	public static void fillArray(int arr[]){
    	int size = arr.length;
    	for (int i = 0; i < size; i++){
        	arr[i] = 0;
    	}
  	}

  	public static void printArray(int arr[]){
    	int size = arr.length;
    	for (int i = 0; i < size; i++){
        	System.out.print(arr[i]);
    	}
    	System.out.print("\n");
  	}

  	public static void complement(int arr[]){
    	int size = arr.length;
    	for (int i = 0; i < size; i++){
        	arr[i] ^= 1;
    	}
  	}

  	public static void reverse(int arr[]){
    	int size = arr.length;
    	int tmp = 0;
    	for (int i = 0; i < Math.floor(size/2); i++){
        	tmp = arr[i];
        	arr[i] = arr[size-1-i];
        	arr[size-1-i] = tmp;
    	}
  	}

  	public static int query(int i){

    	if(i != -1){
      	System.out.print(i+1);
    	}
    	else{
      	System.out.print("1");
    	}
    	System.out.print("\n");
    	count++;
    	int input = sc.nextInt();
    	sc.nextLine();
    	return input;
  	}

  	public static void checkCorrect(int arr[]){
    	printArray(arr);
    	String input = sc.nextLine();

    	if (!input.equals("Y"))
      	System.exit(0);

  	}

  	public static void esab_atad(int B){
    	int result[] = new int[B];
    	int complement_i = -1;
    	int reverse_i = -1;
    	int complement_res = 0;
    	int reverse_res = 0;
    	count = 0;
    	fillArray(result);
    	for (int i = 0; i < Math.floor(B / 2); i++){
      	if(count != 0 && count%10 == 0){
        	complement_res = query(complement_i);
        	reverse_res = query(reverse_i);

        	if ( complement_i != -1 && (result[complement_i]^complement_res) == 1)
          	complement(result);

        	if ( reverse_i != -1 && (result[reverse_i]^reverse_res) == 1
	)
          	reverse(result);
      	}

      	result[i] = query(i);

      	result[B-i-1] = query(B-i-1);

      	if( result[i] == result[B-i-1])
        	complement_i = i;
      	else
        	reverse_i = i;
    	}
    	checkCorrect(result);
  	}

  	public static void main(String[] args) {
    	int T;
    	int B;
    	String input;
    	String[] inputSplit;
    	input = sc.nextLine();
    	inputSplit = input.split(" ");

    	T = Integer.parseInt(inputSplit[0]);
    	B = Integer.parseInt(inputSplit[1]);

    	for (int i = 0; i < T; i++){
      	esab_atad(B);
    	}
  	}
}
